### PR TITLE
Add viewport documentation breadcrumbs

### PR DESCRIPTION
**Pull request by nscobie**
## Description

Backfilling documentation breadcrumbs from debugging https://github.com/flutter/flutter/issues/61631

## Related Issues

N/A

## Tests

N/A

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [ ] All existing and new tests are passing.
- [ ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 1]
```diff
@@ -188,6 +188,19 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
        _cacheExtentStyle = cacheExtentStyle,
        _clipBehavior = clipBehavior;
 
+  /// Report the semantics of this node, for example for accessibility purposes.
+  ///
+  /// Adds [RenderViewport.useTwoPaneSemantics] to support children using
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:193]: this comment is a bit confusing, can you explain more?

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 2]
```diff
@@ -1188,6 +1205,10 @@ class RenderViewport extends RenderViewportBase<SliverPhysicalContainerParentDat
   /// The semantic scrolling actions and the [SemanticsNode]s of scrollable
   /// children will be attached to the inner node, which itself is a child of
   /// the outer node.
+  ///
+  /// See also:
+  ///
+  /// * [RenderViewportBase.describeSemanticsConfiguration] which makes use of this.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:1211]: This should explain how it uses it

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 3]
```diff
@@ -1040,6 +1053,10 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
   ///
   /// The `duration` parameter can be set to a non-zero value to animate the
   /// target object into the viewport with an animation defined by `curve`.
+  ///
+  /// See also:
+  ///
+  /// * [RenderObject.showOnScreen]
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:1059]: this should also add why people would want to look at this

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 4]
```diff
@@ -188,6 +188,19 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
...
+  /// [RenderViewport.excludeFromScrolling].
+  ///
+  /// This method should be overridden by subclasses that have interesting
+  /// semantic information, but `super.describeSemanticsConfiguration(config)`
+  /// should likely be called to ensure [RenderViewport.useTwoPaneSemantics] is
+  /// still added.
+  ///
+  /// See also:
+  ///
+  /// * [RenderObject.describeSemanticsConfiguration] for important details.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:203]: missing ':'.
This should talk a little bit more details

File: packages/flutter/lib/src/rendering/object.dart [hunk 5]
```diff
@@ -3805,6 +3813,12 @@ class _SemanticsGeometry {
   /// Value for [SemanticsNode.rect].
   Rect get rect => _rect;
 
+  /// Computes values, ensuring `rect` is properly bounded by ancestor clipping rects.
+  ///
+  /// See also:
+  ///
+  /// * [RenderObject.describeSemanticsClip]
+  /// * [RenderObject.describeApproximatePaintClip]
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/object.dart:3821]: This should also include why we want to look at this

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/object.dart:3821]: same here and everywhere

**Comment by goderbauer** [comment]
PR triage: @nscobie any chance you can follow up on the comments?

File: packages/flutter/lib/src/rendering/viewport.dart (hunk 1 above)
Comment (nscobie) [reply packages/flutter/lib/src/rendering/viewport.dart]: Tried to clean it up a bit, let me know what you think. It was meant to mirror the super method's documentation: [RenderObject.describeSemanticsConfiguration](https://api.flutter.dev/flutter/rendering/RenderObject/describeSemanticsConfiguration.html).

**Comment by nscobie** [comment]
@chunhtai I appreciate the review! I've added a bit more context to each "see also" reference, and tried to clarify the comment about `describeSemanticsConfiguration`. I don't mind removing anything here that just doesn't seem like helpful detail, but I think these links would have helped me personally when learning how things interacted.

@goderbauer apologies for the delay, the start of the semester has given me less time than I'd like to volunteer on this.

File: packages/flutter/lib/src/rendering/object.dart [hunk 6]
```diff
@@ -2421,6 +2421,11 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   /// semantics tree to implement implicit accessibility scrolling on iOS where
   /// the viewport scrolls implicitly when moving the accessibility focus from
   /// a the last visible node in the viewport to the first hidden one.
+  ///
+  /// See also:
+  ///
+  /// * [RenderViewportBase.cacheExtent], used by viewports to extend their
+  /// semantics clip beyond their approximate paint clip.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/object.dart:2428]: nit: indent

File: packages/flutter/lib/src/rendering/object.dart [hunk 7]
```diff
@@ -2906,6 +2911,11 @@ abstract class RenderObject extends AbstractNode with DiagnosticableTreeMixin im
   ///
   /// The `duration` parameter can be set to a non-zero value to bring the
   /// target object on screen in an animation defined by `curve`.
+  ///
+  /// See also:
+  ///
+  /// * [RenderViewportBase.showInViewport], which this method typically
+  /// delegates to.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/object.dart:2918]: indent

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 8]
```diff
@@ -188,6 +188,21 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
...
+  ///
+  /// This method should be overridden by subclasses that have interesting
+  /// semantic information. Overriding subclasses should call
+  /// `super.describeSemanticsConfiguration(config)` to ensure
+  /// [RenderViewport.useTwoPaneSemantics] is still added to `config`.
+  ///
+  /// See also:
+  ///
+  /// * [RenderObject.describeSemanticsConfiguration], for important
+  /// details about not mutating a [SemanticsConfiguration] out of context.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:205]: indent

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 9]
```diff
@@ -1040,6 +1055,10 @@ abstract class RenderViewportBase<ParentDataClass extends ContainerParentDataMix
   ///
   /// The `duration` parameter can be set to a non-zero value to animate the
   /// target object into the viewport with an animation defined by `curve`.
+  ///
+  /// See also:
+  ///
+  /// * [RenderObject.showOnScreen], commonly used to wrap `showInViewport`.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:1061]: It seems like only two classes(RenderViewportBase, SingleChildScrollView) are using this in their show on screen. Let's list two of them separately here

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 10]
```diff
@@ -1188,6 +1207,11 @@ class RenderViewport extends RenderViewportBase<SliverPhysicalContainerParentDat
   /// The semantic scrolling actions and the [SemanticsNode]s of scrollable
   /// children will be attached to the inner node, which itself is a child of
   /// the outer node.
+  ///
+  /// See also:
+  ///
+  /// * [RenderViewportBase.describeSemanticsConfiguration], which adds this
+  /// tag to its [SemanticsConfiguration] as described above.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:1214]: indent

File: packages/flutter/lib/src/rendering/object.dart (hunk 7 above)
Comment (chunhtai) [reply packages/flutter/lib/src/rendering/object.dart:2918]: It maybe a little bit inaccurate, we do not override this method often, it just so happens that we uses this method for those occasion.

**Review by chunhtai — COMMENTED** [review]
These changes are great, they add more explanation to a complicated system.

File: packages/flutter/lib/src/rendering/object.dart (hunk 6 above)
Comment (nscobie) [reply packages/flutter/lib/src/rendering/object.dart:2428]: Just to be clear, should the indent on the second line match the position of the `[` on the first line, or be indented even further?

Comment (chunhtai) [reply packages/flutter/lib/src/rendering/object.dart:2428]: it should match the position of `[`

**Comment by nscobie** [comment]
@googlebot I signed it!

File: packages/flutter/lib/src/rendering/viewport.dart [hunk 11]
```diff
@@ -1188,6 +1209,11 @@ class RenderViewport extends RenderViewportBase<SliverPhysicalContainerParentDat
   /// The semantic scrolling actions and the [SemanticsNode]s of scrollable
   /// children will be attached to the inner node, which itself is a child of
   /// the outer node.
+  ///
+  /// See also:
+  ///
+  /// * [RenderViewportBase.describeSemanticsConfiguration], which adds this
+  ///   tag to its [SemanticsConfiguration] as described above.
```
Comment (chunhtai) [inline packages/flutter/lib/src/rendering/viewport.dart:1216]: nit, let's remove the `as described above.` It seems to not add much information

**Review by chunhtai — APPROVED** [review]
LGTM

**Comment by nscobie** [comment]
> LGTM

👍 Thanks again!

Heads up, I no longer have write access, so you'll need to merge for me (assuming tests pass).

**Comment by chunhtai** [comment]
Hi @nscobie, the doc presubmit check is failing.

**Comment by nscobie** [comment]
Huh, looks like a crash to me. @chunhtai would you mind initiating a re-run?

```
dartdoc:stdout: Generating docs for library webdriver.sync_io from package:webdriver/sync_io.dart...
dartdoc:stderr: found 0 warnings and 1 error
dartdoc:stdout: Documented 112 public libraries in 730.7 seconds
dartdoc:stderr:
dartdoc:stderr: dartdoc failed: dartdoc encountered 1 errors while processing..
dartdoc:stderr: #0      Dartdoc.generateDocs (package:dartdoc/dartdoc.dart:225:9)
dartdoc:stderr: <asynchronous suspension>
dartdoc:stderr: #1      Dartdoc.executeGuarded.<anonymous closure> (package:dartdoc/dartdoc.dart:485:15)
dartdoc:stderr: #2      _rootRun (dart:async/zone.dart:1190:13)
dartdoc:stderr: #3      _CustomZone.run (dart:async/zone.dart:1093:19)
dartdoc:stderr: #4      _runZoned (dart:async/zone.dart:1630:10)
dartdoc:stderr: #5      runZonedGuarded (dart:async/zone.dart:1618:12)
dartdoc:stderr: #6      Dartdoc.executeGuarded (package:dartdoc/dartdoc.dart:483:5)
dartdoc:stderr: #7      main (file:///root/.pub-cache/hosted/pub.dartlang.org/dartdoc-0.32.3/bin/dartdoc.dart:24:11)
dartdoc:stderr: <asynchronous suspension>
dartdoc:stderr: #8      _startIsolate.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:299:32)
dartdoc:stderr: #9      _RawReceivePortImpl._handleMessage (dart:isolate-patch/isolate_patch.dart:168:12)
dartdoc:stderr:
������������
```

**Comment by chunhtai** [comment]
Hi @nscobie it is not a crash https://api.cirrus-ci.com/v1/task/5909429256192000/logs/main.log
```
dartdoc:stdout: Generating docs for library material from package:flutter/material.dart...
dartdoc:stdout: Generating docs for library painting from package:flutter/painting.dart...
dartdoc:stdout: Generating docs for library physics from package:flutter/physics.dart...
dartdoc:stdout: Generating docs for library rendering from package:flutter/rendering.dart...
dartdoc:stderr:   error: unresolved doc reference [SingleChildScrollView.showOnScreen], from rendering.RenderViewportBase.showInViewport: (file:///tmp/flutter%20sdk/packages/flutter/lib/src/rendering/viewport.dart:1064:15)
dartdoc:stdout: Generating docs for library scheduler from package:flutter/scheduler.dart...
dartdoc:stdout: Generating docs for library semantics from package:flutter/semantics.dart...
```

**Comment by nscobie** [comment]
Ah thanks for pointing that out, I should have looked at the full log.

It's because I overlooked that `SingleChildScrollView` actually relies on `_RenderSingleChildViewport` to override `showOnScreen`. I'll update the docs to reflect that.

**Comment by nscobie** [comment]
CI looks happy now 🎉

**Pull request merged**